weasyl/favorite.py

```diff
@@ -195,40 +195,78 @@ def insert(userid, submitid=None, charid=None, journalid=None):
     elif ignoreuser.check(query[0], userid):
         raise WeasylError("contentOwnerIgnoredYou")
 
-    insert_result = d.engine.execute(
-        'INSERT INTO favorite (userid, targetid, type, unixtime) '
-        'VALUES (%(user)s, %(target)s, %(type)s, %(now)s) '
-        'ON CONFLICT DO NOTHING',
-        user=userid,
-        target=d.get_targetid(submitid, charid, journalid),
-        type='s' if submitid else 'f' if charid else 'j',
-        now=d.get_time())
-
-    if insert_result.rowcount == 0:
-        return
-
-    # create a list of users to notify
-    notified = set(collection.find_owners(submitid))
-
-    # conditions under which "other" should be notified
-    def can_notify(other):
-        other_jsonb = d.get_profile_settings(other)
-        allow_notify = other_jsonb.allow_collection_notifs
-        not_ignored = not ignoreuser.check(other, userid)
-        return allow_notify and not_ignored
-    notified = set(filter(can_notify, notified))
-    # always notify for own content
-    notified.add(query[0])
-
-    for other in notified:
-        welcome.favorite_insert(userid, submitid=submitid, charid=charid, journalid=journalid, otherid=other)
+    notified = []
+
+    def insert_transaction(db):
+        insert_result = db.execute(
+            'INSERT INTO favorite (userid, targetid, type, unixtime) '
+            'VALUES (%(user)s, %(target)s, %(type)s, %(now)s) '
+            'ON CONFLICT DO NOTHING',
+            user=userid,
+            target=d.get_targetid(submitid, charid, journalid),
+            type='s' if submitid else 'f' if charid else 'j',
+            now=d.get_time())
+
+        if insert_result.rowcount == 0:
+            return
+
+        if submitid:
+            db.execute(
+                """
+                UPDATE submission SET
+                    favorites = (
+                        CASE
+                            WHEN favorites IS NULL THEN
+                                (SELECT count(*) FROM favorite WHERE type = 's' AND targetid = %(target)s)
+                            ELSE
+                                favorites + 1
+                        END
+                    )
+                    WHERE submitid = %(target)s
+                """,
+                target=submitid,
+            )
+
+        if not notified:
```

Comment:
`notified.extend()` and `notified.append()` below fill the list. This is weird and hard to read, but I couldn’t find a way around it, since Python 2 doesn’t support `nonlocal`.